File: scala-server/app/com/bolour/eagerwords/common/domain/DeviceType.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.common.domain

object DeviceType {

  val MouseString = "Mouse"
  val TouchString = "Touch"

  sealed abstract class DeviceType

  object MouseDevice extends DeviceType {
    override def toString = MouseString
  }

  object TouchDevice extends DeviceType {
    override def toString = TouchString
  }

  def fromString(asString: String): DeviceType = {
    asString match {
      case MouseString => DeviceType.MouseDevice
      case TouchString => DeviceType.TouchDevice
      case _ => throw new RuntimeException(s"unrecognized device type: ${asString}")
    }
  }
}

File: haskell-server/src/EagerWords/Common/Domain/DeviceType.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module EagerWords.Common.Domain.DeviceType (
    DeviceType (..)
) where

import GHC.Generics
import Data.Aeson (FromJSON, ToJSON)
import Control.DeepSeq (NFData)
import qualified Data.Aeson as Aeson

-- | Types of input devices.
data DeviceType = Mouse | Touch
  deriving (Show, Read, Eq, NFData, Generic)

instance FromJSON DeviceType
instance ToJSON DeviceType






